model-optimizer/extensions/middle/LayoutChangeForEinsum.py

```diff
@@ -19,7 +21,13 @@ class LayoutChangeForEinsum(BackReplacementPattern):
     """
     enabled = True
     force_shape_inference = True
-    graph_condition = [lambda graph: graph.graph['fw'] == 'tf']
+    graph_condition = [lambda graph: graph.graph['layout'] == 'NHWC']
+
+    def run_after(self):
+        return [InsertLayoutPropagationTranspose]
+
+    def run_before(self):
+        return []
```

Comment:
You should specify some valid dependency here. Otherwise your transformation will be a dangling node in the dependency graph and may be executed even during the back phase.

Reply:
Great catch to make it reliable. `LayoutChangeForEinsum` can be executed even right after `MarkSubgraphWithCorrectLayout` as I did now because `correct_layout` attribute is set within `MarkSubgraphWithCorrectLayout`. Also, `InsertLayoutPropagationTranspose` does not have "before" dependency that disturbed to setup "before" dependency in the previous version.
Done.